cloud/amazon/datapipeline.py

```diff
@@ -0,0 +1,470 @@
+
+
+def pipeline_field(client, dp_id, field):
+    """Return a pipeline field from the pipeline description.
+
+    The available fields are listed in describe_pipelines output.
+
+    :param object client: boto3 datapipeline client
+    :param string dp_id: pipeline id
+    :param string field: pipeline description field
+    :returns: pipeline field information
+
+    """
+    dp_description = pipeline_description(client, dp_id)
+    for field_key in dp_description['pipelineDescriptionList'][0]['fields']:
+        if field_key['key'] == field:
+            return field_key['stringValue']
+    else:
+        raise KeyError("Field key {} not found!".format(field))
+
+pipeline_status = partial(pipeline_field, field='@pipelineState')
+pipeline_uniqueid = partial(pipeline_field, field='uniqueId')
+
+
+def run_with_timeout(timeout, func, *func_args, **func_kwargs):
+        """Run func with the provided args and kwargs, and wait utill
+        timeout for truthy return value
+
+        :param int timeout: time to wait for status
+        :param function func: function to run, should return True or False
+        :param args func_args: function args to pass to func
+        :param kwargs func_kwargs: function key word args
+        :returns: True if func returns truthy within timeout
+        :raises: TimeOutException
+
+        """
+        start = datetime.datetime.now()
+        end = start + datetime.timedelta(seconds=timeout)
+
+        while datetime.datetime.now() < end:
+            if func(*func_args, **func_kwargs):
+                return True
+            else:
+                # check every 10s
+                time.sleep(10)
+
+        raise TimeOutException
+
+
+def check_dp_exists(client, dp_id):
+    """Check if datapipeline exists
+
+    :param object client: boto3 datapipeline client
+    :param string dp_id: pipeline id
+    :returns: True or False
+
+    """
+    try:
+        # pipeline_description raises DataPipelineNotFound
+        if pipeline_description(client, dp_id):
+            return True
+        else:
+            return False
+    except DataPipelineNotFound:
+        return False
+
+
+def check_dp_status(client, dp_id, status):
+    """Checks if datapipeline matches states in status list
+
+    :param object client: boto3 datapipeline client
+    :param string dp_id: pipeline id
+    :param list status: list of states to check against
+    :returns: True or False
+
+    """
+    assert type(status) == list
+    if pipeline_status(client, dp_id) in status:
+        return True
+    else:
+        return False
+
+
+def pipeline_status_timeout(client, dp_id, status, timeout):
+    args = (client, dp_id, status)
+    try:
+        return run_with_timeout(timeout, check_dp_status, *args)
+    except TimeOutException:
+        raise
+
+
+def pipeline_exists_timeout(client, dp_id, timeout):
+    args = (client, dp_id)
+    try:
+        return run_with_timeout(timeout, check_dp_exists, *args)
```

Comment:
You should use the boto pipeline get_waiter() method and boto3 waiter rather than DIY http://boto3.readthedocs.io/en/latest/guide/clients.html